save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/TopBar.kt

```diff
@@ -54,6 +54,148 @@ external interface TopBarProps : PropsWithChildren {
     var userInfo: UserInfo?
 }
 
+/**
+ * The class for analyzing url address and creating links in topBar
+ */
+class TopBarUrl(href: String) {
+    /**
+     * currentPath is the link that we put in buttons
+     */
+    var currentPath = "#"
+    private var circumstance: SituationUrlClassification = SituationUrlClassification.KEYWORD_PROCESS
+    private var processLastSegments = 0
+    private val sizeUrlSegments: Int
+
+    init {
+        sizeUrlSegments = href.split("/").size
+        findExclude(href)
+    }
+
+    /**
+     * The function is called to specify the link address in the button before creating the button itself
+     *
+     * @param pathPart is an appended suffix to an already existing [currentPath]
+     */
+    fun changeUrlBeforeButton(pathPart: String) {
+        currentPath = when (circumstance) {
+            SituationUrlClassification.PROJECT_OR_ORGANIZATION -> "#/${FrontendRoutes.PROJECTS.path}"
+            SituationUrlClassification.ARCHIVE -> "#/${FrontendRoutes.AWESOME_BENCHMARKS.path}"
+            SituationUrlClassification.DETAILS, SituationUrlClassification.EXECUTION -> if (pathPart == "execution") currentPath else mergeUrls(pathPart)
+            else -> mergeUrls(pathPart)
+        }
+    }
+
+    /**
+     * The function is called to specify the link address in the button after creating the button itself
+     *
+     * @param pathPart is an appended suffix to an already existing [currentPath]
+     */
+    fun changeUrlAfterButton(pathPart: String) {
+        fixCurrentPathAfter(pathPart)
+        fixExcludeAfter("\\d+".toRegex().matches(pathPart))
+    }
+
+    /** The function set a flag whether to create this button or not
+     *
+     * @param index
+     */
+    fun isCreateButton(index: Int) = when (circumstance) {
+        SituationUrlClassification.KEYWORD_PROCESS_LAST_SEGMENTS -> index >= sizeUrlSegments - 1 - processLastSegments
+        SituationUrlClassification.KEYWORD_NOT_PROCESS -> false
+        else -> true
+    }
+
+    /** The function
+     *
+     * @param href
+     */
+    private fun findExclude(href: String) {
+        circumstance = if (href.contains(OrganizationMenuBar.regexForUrlClassification) || href.contains(ProjectMenuBar.regexForUrlClassification)) {
+            SituationUrlClassification.PROJECT_OR_ORGANIZATION
+        } else if (href.contains(BenchmarkCategoryEnum.regexForUrlClassification)) {
+            SituationUrlClassification.ARCHIVE
+        } else if (href.contains(Regex("/[^/]+/[^/]+/history/execution/[1234567890]+/details"))) {
+            SituationUrlClassification.DETAILS
```

Comment:
This regex looks good to be stored in enum
But if you would like to fix it, you need to change it all
Better ask @petertrr if it needs to be changed or it is fine now

Reply:
Suggestion: we have `TabMenuBar`, add a new field there with type `SituationUrlClassification` and then it can be refactored to something like:
```kotlin
    listOf<TabMenuBar<*>>(OrganizationMenuBar, BenchmarkCategoryEnum, ...)
        .firstOrNull { "".contains(it.regexForUrlClassification) }
        ?.situationUrlClassification
        ?: SituationUrlClassification.KEYWORD_PROCESS
```

Reply:
Done